pymod/ams.py

```diff
@@ -170,17 +172,76 @@ class ArgoMessagingService(AmsHttpRequests):
        calls that are wrapped in series of methods. Class is entry point for
        client code.
     """
-    def __init__(self, endpoint, token="", project=""):
+    def __init__(self, endpoint, token="", project="", cert="", key=""):
         super(ArgoMessagingService, self).__init__()
+        self.token = ""
         self.endpoint = endpoint
-        self.token = token
         self.project = project
+        self.assign_token(token, cert, key)
         self.pullopts = {"maxMessages": "1",
                          "returnImmediately": "false"}
         # Containers for topic and subscription objects
         self.topics = dict()
         self.subs = dict()
 
+    def assign_token(self, token, cert, key):
+        """
+        Assign a token to the ams object
+
+        Args:
+            token(str): a valid ams token
+            cert(str): a path to a valid certificate file
+            key(str): a path to the associated key file for the provided certificate
+        """
+
+        # check if a token has been provided
+        if token != "":
+            self.token = token
+            return
+
+        try:
+            # otherwise use the provided certificate to retrieve it
+            self.token = self.auth_via_cert(cert, key)
+        except AmsServiceException as e:
+            if e.message["error"] == 'While trying the [auth_x509]: No certificate provided.':
+                e.message["error"] += "No token provided"
```

Comment:
Let get guess, if the error is While trying the [auth_x509]: No certificate provided. we give to the user the error "No Token Provided"  ? am I missing something or is this error simply wrong ?

Reply:
Also we need to make sure we keep backwards compatibility i.e if not cert is supplied the library works as before.

Reply:
@kkoumantaros 
1) For the error message: The message is appended to the already existing one, thus, the result should be 'While trying the [auth_x509]: No certificate provided.No token provided. This can also be seen from the test files.

2) The initialization takes either a token or a cert/key. All of them are optional parameters.If you initialize an ams object providing a token, it will work just as before.